unit_tests/recovery/ut_database.cpp

```diff
@@ -0,0 +1,363 @@
+      const auto &page = m_pages.at (page_index);
+      return page->generate_changes (a_database_recovery, a_db_global_values);
+    }
+  else
+    {
+      if (add_or_update_page == ut_database_values_generator::add_or_update::ADD
+	  || m_pages.size () == 0)
+	{
+	  // add new page and generate log entry
+	  // akin to 'extend volume' operation
+	  const auto lsa_log_id = a_db_global_values.increment_and_get_lsa_log_id ();
+	  const int millis = a_db_global_values.rand_duration_in_millis ();
+	  ux_ut_redo_job_impl job
+	  {
+	    new ut_redo_job_impl (a_database_recovery, ut_redo_job_impl::job_type::NEW_PAGE,
+	    lsa_log_id, {static_cast<short> (m_pages.size ()), m_volid}, millis)
+	  };
+	  add_new_page (m_pages);
+	  return job;
+	}
+      else
+	{
+	  assert (false);
+	}
+    }
+  return nullptr;
+}
+
+void ut_volume::apply_changes (ux_ut_redo_job_impl &&a_job)
+{
+  // NOTE: although function is to be called from different threads, the function
+  // is intentionally not locked as the intention is to check proper synchronization in
+  // the redo log apply algorithm
+  if (a_job->is_volume_creation ())
+    {
+      assert (false);
+    }
+  else
+    {
+      if (a_job->is_page_creation ())
+	{
+	  std::lock_guard<std::mutex> lock (m_apply_changes_mtx);
+	  const ux_ut_page &new_page = add_new_page (m_pages);
+	  assert (new_page->get_vpid ().volid == a_job->get_vpid ().volid);
+	}
+      else
+	{
+	  std::lock_guard<std::mutex> lock (m_apply_changes_mtx);
+	  ux_ut_page &page = m_pages.at (a_job->get_vpid ().pageid);
+	  page->apply_changes (std::move (a_job));
+	}
+    }
+}
+
+bool ut_volume::operator== (const ut_volume &that) const
+{
+  if (m_volid != that.m_volid || m_pages.size () != that.m_pages.size ())
+    {
+      return false;
+    }
+  auto this_pages_it = m_pages.cbegin ();
+  auto that_pages_it = that.m_pages.cbegin ();
+  while (this_pages_it != m_pages.cend () && that_pages_it != that.m_pages.cend ())
+    {
+      if (! (**this_pages_it == **that_pages_it))
+	{
+	  return false;
+	}
+      ++this_pages_it;
+      ++that_pages_it;
+    }
+  return true;
+}
+
+const ux_ut_page &ut_volume::add_new_page (std::vector<ux_ut_page> &a_pages)
+{
+  ux_ut_page page { new ut_page (m_volid, a_pages.size ()) };
+  //page->initialize (m_volid, a_pages.size ());
+  a_pages.push_back (std::move (page));
+  return *a_pages.rbegin ();
+}
+
+/*******************************************************
+ * ut_database
+ *******************************************************/
+
+ut_database::ut_database (const ut_database_config &a_database_config)
+  : m_database_config (a_database_config)
+{
+  // start without any volumes
+}
+
+ux_ut_redo_job_impl ut_database::generate_changes (ut_database &a_database_recovery,
+    ut_database_values_generator &a_db_global_values)
+{
+  //const int update_or_add_volume = rand_update_or_add_volume();
+  const auto add_or_update_volume = a_db_global_values.rand_add_or_update_volume (m_volumes.size ());
```

Comment:
A suggestion to avoid checking m_volumes.size() again.
```suggestion
  auto add_or_update_volume = ut_database_values_generator::add_or_update::ADD;
  if (m_volumes.size () > 0)
    {
      add_or_update_volume = a_db_global_values.rand_add_or_update_volume (m_volumes.size ());
    }
```

Reply:
Done